File: file/src/main/scala/farjs/file/popups/EncodingsPopup.scala
package farjs.file.popups

import farjs.file.Encoding
import farjs.ui.popup._
import scommons.react._
import scommons.react.hooks._

import scala.scalajs.js

object EncodingsPopup extends FunctionComponent[EncodingsPopupProps] {

  private[file] var listPopup: ReactClass = ListPopup
  
  protected def render(compProps: Props): ReactElement = {
    val props = compProps.plain
    val (maybeItems, setItems) = useState(Option.empty[js.Array[String]])

    useLayoutEffect({ () =>
      setItems(Some(js.Array(Encoding.encodings: _*)))
      ()
    }, Nil)

    maybeItems.map { items =>
      <(listPopup)(^.plain := ListPopupProps(
        title = "Encodings",
        items = items,
        onAction = { index =>
          val enc = items(index)
          if (enc != props.encoding) {
            props.onApply(enc)
          }
          props.onClose()
        },
        onClose = props.onClose,
        selected = math.max(items.indexOf(props.encoding), 0)
      ))()
    }.orNull
  }
}

File: file/popups/EncodingsPopup.mjs
import React, { useLayoutEffect, useState } from "react";
import ListPopup from "@farjs/ui/popup/ListPopup.mjs";
import Encoding from "../Encoding.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly encoding: string;
 *  onApply(encoding: string): void;
 *  onClose(): void;
 * }} EncodingsPopupProps
 */

/**
 * @param {EncodingsPopupProps} props
 */
const EncodingsPopup = (props) => {
  const { listPopup } = EncodingsPopup;

  const [maybeItems, setItems] = useState(
    /** @type {string[] | undefined} */ (undefined)
  );

  useLayoutEffect(() => {
    setItems(Encoding.encodings);
  }, []);

  if (maybeItems !== undefined) {
    return h(listPopup, {
      title: "Encodings",
      items: maybeItems,
      onAction: (index) => {
        const enc = maybeItems[index];
        if (enc !== props.encoding) {
          props.onApply(enc);
        }
        props.onClose();
      },
      onClose: props.onClose,
      selected: Math.max(maybeItems.indexOf(props.encoding), 0),
    });
  }

  return null;
};

EncodingsPopup.displayName = "EncodingsPopup";
EncodingsPopup.listPopup = ListPopup;

export default EncodingsPopup;
